src/parser/SparqlParser.cpp

```diff
@@ -188,43 +188,74 @@ void SparqlParser::parseWhere(const string& str, ParsedQuery& query,
           inner.substr(k, 8) == "optional") {
         // find opening and closing brackets of optional part
         size_t ob = inner.find('{', k);
-        size_t cb = ob;
-        int depth = 0;
-        size_t i;
-        for (i = ob + 1; i < inner.size(); i++) {
-          if (inner[i] == '{') {
-            depth++;
-          }
-          if (inner[i] == '}') {
-            if (depth == 0) {
-              cb = i;
-              break;
-            } else {
-              depth--;
-            }
-          }
-        }
-        if (i == inner.size()) {
-          if (depth == 0) {
-            throw ParseException(
-                "Need curly braces in optional"
-                "clause.");
-          } else {
-            throw ParseException("Unbalanced curly braces.");
-          }
-        }
-        currentPattern->_children.push_back(new ParsedQuery::GraphPattern());
-        currentPattern->_children.back()->_optional = true;
-        currentPattern->_children.back()->_id = query._numGraphPatterns;
+        size_t cb = ad_utility::findClosingBracket(inner, ob, '{', '}');
+        currentPattern->_children.push_back(
+            new ParsedQuery::GraphPatternOperation(
```

Comment:
where are these freed? Can't find it in `GraphPattern` and that has `std::vector` of raw pointers not `std::shared_ptr`

Reply:
The GraphPattern destructor deletes them.